<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<!-- saved from url=(0069)http://www.broadinstitute.org/~orzuk/matlab/matlab_utilities_template.html -->
<HTML xmlns="http://www.w3.org/1999/xhtml"><HEAD><META http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<!--<LINK rel="SHORTCUT ICON" href="orzuk.ico">-->
<LINK rel="stylesheet" href="http://www.broadinstitute.org/~orzuk/orzuk.css" type="text/css"> 
<TITLE>Matlab bioinfo utilities</TITLE>
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-7650195-3']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
</HEAD><BODY>


<DIV class="navbar"> 
<a href="http://www.broadinstitute.org/~orzuk/matlab/libs/matlab_utils.html">libs</a>  :
</DIV> 
  <DIV class="sep">
    <HR>
  </DIV>

  <DIV class="main">
    <H2>Matlab bioinfo utilities</H2>

    <UL>

      <LI> Matlab bioinfo utilities (see documentation and examples within the functions)</LI>

    </UL>

<!--    <p>Changes are now logged in the <a href="http://home.online.no/%7Epjacklam/matlab/software/util/statutil/ChangeLog">ChangeLog</a> file.</p>--><!-- Begin m-file index -->
    <!-- This table was automatically generated -->
    <!-- by updhtml.pl on Sat Feb 21 14:26:51 2004 -->

    <TABLE summary="M-file index" border="0" cellpadding="0" cellspacing="0">
      <TBODY><TR>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/AgilentBasicToCompleteProbesFormat.m" name="AgilentBasicToCompleteProbesFormat" id="AgilentBasicToCompleteProbesFormat">AgilentBasicToCompleteProbesFormat.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Converts a basic probes format file to a complete format for agilent eArray. </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/Assign24MammalsBasicGlobalConstants.m" name="Assign24MammalsBasicGlobalConstants" id="Assign24MammalsBasicGlobalConstants">Assign24MammalsBasicGlobalConstants.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>All kinds of constants and global variables for the 24 mammals project.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/Assign24MammalsGlobalConstants.m" name="Assign24MammalsGlobalConstants" id="Assign24MammalsGlobalConstants">Assign24MammalsGlobalConstants.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>All kinds of constants and global variables for the 24 mammals project</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/AssignGeneralConstants.m" name="AssignGeneralConstants" id="AssignGeneralConstants">AssignGeneralConstants.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>function AssignGeneralConstants()</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/AssignPilotConstants.m" name="AssignPilotConstants" id="AssignPilotConstants">AssignPilotConstants.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>TERNAL = 0; TOPHAT = 1;</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/AssignStatsConstants.m" name="AssignStatsConstants" id="AssignStatsConstants">AssignStatsConstants.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>AssignStatsConstants()</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/ComputeGeneMutationRates.m" name="ComputeGeneMutationRates" id="ComputeGeneMutationRates">ComputeGeneMutationRates.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Calculate total mutation rates for different types of coding mutations for the human genome</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/CountGenomeKmers.m" name="CountGenomeKmers" id="CountGenomeKmers">CountGenomeKmers.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Go over the WHOLE genome and count the occurances of each kmer.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/CountGenomicRegionsKmers.m" name="CountGenomicRegionsKmers" id="CountGenomicRegionsKmers">CountGenomicRegionsKmers.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Count kmers in a set of coordinates given</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/FastaGenomeToMatFiles.m" name="FastaGenomeToMatFiles" id="FastaGenomeToMatFiles">FastaGenomeToMatFiles.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert whole genome chromsomes in fasta format into binary .mat files. </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/FormatChangeGenomicRegions.m" name="FormatChangeGenomicRegions" id="FormatChangeGenomicRegions">FormatChangeGenomicRegions.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert .txt format chr5:start-end to numeric: 5 start end</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/GenomicRegionsFiletoMat.m" name="GenomicRegionsFiletoMat" id="GenomicRegionsFiletoMat">GenomicRegionsFiletoMat.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Parse the genomic regions file obtained from Mitch,</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/GenomicRegionsKmersConservation.m" name="GenomicRegionsKmersConservation" id="GenomicRegionsKmersConservation">GenomicRegionsKmersConservation.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Look at all genomic regions, and determine conserved k-mers for each of them</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/GenomicRegionsStatistics.m" name="GenomicRegionsStatistics" id="GenomicRegionsStatistics">GenomicRegionsStatistics.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute some statistics on the genomic regions</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/GetAllConnectedSubTrees.m" name="GetAllConnectedSubTrees" id="GetAllConnectedSubTrees">GetAllConnectedSubTrees.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Extract a sub-tree from a subset of leaves of a phylogenetic tree</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/GetBranchLength.m" name="GetBranchLength" id="GetBranchLength">GetBranchLength.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute the total branch length of a subset of leaves of a phylogenetic tree</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/GetGenomeLength.m" name="GetGenomeLength" id="GetGenomeLength">GetGenomeLength.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Gets the length of all the chromosomes in a given genome. </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/GetRandomGenomicRegions.m" name="GetRandomGenomicRegions" id="GetRandomGenomicRegions">GetRandomGenomicRegions.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Return coordinates to regions selected randomly and uniformly from the genome</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/GetSubTree.m" name="GetSubTree" id="GetSubTree">GetSubTree.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Extract a sub-tree from a subset of leaves of a phylogenetic tree</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/HumanGenomeToMatFiles.m" name="HumanGenomeToMatFiles" id="HumanGenomeToMatFiles">HumanGenomeToMatFiles.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert all the human genome into binary .mat files. </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/MyBaseCount.m" name="MyBaseCount" id="MyBaseCount">MyBaseCount.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Similar to Matlab Bioinformatics Toolbox's 'basecount', </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/PlotGenomicRegions.m" name="PlotGenomicRegions" id="PlotGenomicRegions">PlotGenomicRegions.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Plot the conservation of each type of genomic region </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/ReadBedFileToMat.m" name="ReadBedFileToMat" id="ReadBedFileToMat">ReadBedFileToMat.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Just like ReadWigFileToMat, with another name</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/ReadGeneSetsToMatfile.m" name="ReadGeneSetsToMatfile" id="ReadGeneSetsToMatfile">ReadGeneSetsToMatfile.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert GSEA gmt format gene sets to .mat files </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/ReadWigFileToMat.m" name="ReadWigFileToMat" id="ReadWigFileToMat">ReadWigFileToMat.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Read a wig formatted file into mat file.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/SaveInBedFile.m" name="SaveInBedFile" id="SaveInBedFile">SaveInBedFile.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Save data in .bed format:</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/StratifyToGenomicRegions.m" name="StratifyToGenomicRegions" id="StratifyToGenomicRegions">StratifyToGenomicRegions.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Stratify a set of genomic positions into different genomic regions.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/TreeToDistanceMatrix.m" name="TreeToDistanceMatrix" id="TreeToDistanceMatrix">TreeToDistanceMatrix.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute the distance matrix associated with the tree </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/TwoPWMsToSimilarityMatrix.m" name="TwoPWMsToSimilarityMatrix" id="TwoPWMsToSimilarityMatrix">TwoPWMsToSimilarityMatrix.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Calculate the pairwise similarity matrix between all pairs</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/TwoPWMsToSimilarityVector.m" name="TwoPWMsToSimilarityVector" id="TwoPWMsToSimilarityVector">TwoPWMsToSimilarityVector.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Calculate the pairwise similarity matrix between all pairs</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/TwoTreesToSimilarity.m" name="TwoTreesToSimilarity" id="TwoTreesToSimilarity">TwoTreesToSimilarity.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute similarity/distance between two phylogenetic trees </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/basecount_for_pwm.m" name="basecount_for_pwm" id="basecount_for_pwm">basecount_for_pwm.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Counts nucleotides for a pwm</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/bed_to_wig.m" name="bed_to_wig" id="bed_to_wig">bed_to_wig.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert .bed format (each element is transcript) to .wig (each element an exon)</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/calc_all_sites_scores_from_file.m" name="calc_all_sites_scores_from_file" id="calc_all_sites_scores_from_file">calc_all_sites_scores_from_file.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Scan a set of sequences given as an input file,</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/calc_all_sites_scores_matlab.m" name="calc_all_sites_scores_matlab" id="calc_all_sites_scores_matlab">calc_all_sites_scores_matlab.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>A matlab interface to the c function calc_all_sites_scores.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/chr_num2str.m" name="chr_num2str" id="chr_num2str">chr_num2str.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert number to string with special attention to x,y and m chroms</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/chr_str2num.m" name="chr_str2num" id="chr_str2num">chr_str2num.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert string to number with special attention to x,y and m chroms</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/cluster_to_pwmv2.m" name="cluster_to_pwmv2" id="cluster_to_pwmv2">cluster_to_pwmv2.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Generate a pwm from a set of sequences (cluster)</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/consensus_to_pwm.m" name="consensus_to_pwm" id="consensus_to_pwm">consensus_to_pwm.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert a consensus sequence to a pwm (no derichlet correction)</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/diff_genomic_regions.m" name="diff_genomic_regions" id="diff_genomic_regions">diff_genomic_regions.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Take all regions appearing in one set of genomic regions and not the other</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/draw_many_pwms.m" name="draw_many_pwms" id="draw_many_pwms">draw_many_pwms.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Display many pwms in the same plot</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/draw_phytree.m" name="draw_phytree" id="draw_phytree">draw_phytree.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Draws a phylogenetic tree with weights on the branches, </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/draw_pwm.m" name="draw_pwm" id="draw_pwm">draw_pwm.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Plot a pwm. Similar to Matlab's seqlogo but gives a 'normal' plot window</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/draw_pwm_new.m" name="draw_pwm_new" id="draw_pwm_new">draw_pwm_new.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Here get the height of each column to be according to its information</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/find_best_pwm_matches_in_regions_from_file.m" name="find_best_pwm_matches_in_regions_from_file" id="find_best_pwm_matches_in_regions_from_file">find_best_pwm_matches_in_regions_from_file.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Look for motif hits given coordinates, with input from a file</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/find_best_pwm_matches_in_sequences_pack.m" name="find_best_pwm_matches_in_sequences_pack" id="find_best_pwm_matches_in_sequences_pack">find_best_pwm_matches_in_sequences_pack.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Find matches to position weight matrices in a given set of sequences.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/find_pwms_in_sequences_pack.m" name="find_pwms_in_sequences_pack" id="find_pwms_in_sequences_pack">find_pwms_in_sequences_pack.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Find the matches to position weight matrices in a given set of sequences.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/gc_for_pwm.m" name="gc_for_pwm" id="gc_for_pwm">gc_for_pwm.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute the GC content of a pwm</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/genome_ver_to_organism.m" name="genome_ver_to_organism" id="genome_ver_to_organism">genome_ver_to_organism.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>A simple function telling us for a given genome version what is the</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/get_mammals_model_chr.m" name="get_mammals_model_chr" id="get_mammals_model_chr">get_mammals_model_chr.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Get all chromosome-specific model files</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/get_mutation_types.m" name="get_mutation_types" id="get_mutation_types">get_mutation_types.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute a 64x64 table with all possible mutation types</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/get_scores_above_threshold_Matlab.m" name="get_scores_above_threshold_Matlab" id="get_scores_above_threshold_Matlab">get_scores_above_threshold_Matlab.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Get all the binding sites for a given motif which are</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/intersect_genomic_regions.m" name="intersect_genomic_regions" id="intersect_genomic_regions">intersect_genomic_regions.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Intersect two sets of genomic regions</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/intersect_pwms.m" name="intersect_pwms" id="intersect_pwms">intersect_pwms.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Intersect two sets of pwms </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/iupac_to_degeneracy.m" name="iupac_to_degeneracy" id="iupac_to_degeneracy">iupac_to_degeneracy.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert a iupac code sequence to degeneracy</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/iupac_to_pwm.m" name="iupac_to_pwm" id="iupac_to_pwm">iupac_to_pwm.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert a iupac code sequence to a pwm </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/load_pwms_from_txt_file.m" name="load_pwms_from_txt_file" id="load_pwms_from_txt_file">load_pwms_from_txt_file.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Load a set of pwms in a format convenient for humans to write in txt.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/lowess_two_arrays.m" name="lowess_two_arrays" id="lowess_two_arrays">lowess_two_arrays.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Correct data of two expression vectors using lowess algorithm (by Amit Zeisel)</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/merge_genomic_regions.m" name="merge_genomic_regions" id="merge_genomic_regions">merge_genomic_regions.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Merge (and remove overlap) a set of genomic regions</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/model_tree_read.m" name="model_tree_read" id="model_tree_read">model_tree_read.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Reads a phylogenetic model: this include both tree and rate matrices</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/my_seqlogo.m" name="my_seqlogo" id="my_seqlogo">my_seqlogo.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute seqlogo without displaying, fill in missing letters and normalize</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/my_seqlogov2.m" name="my_seqlogov2" id="my_seqlogov2">my_seqlogov2.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute a weighted average of aligned sequences + pseudocounts to form a pwm</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pack_seqs.m" name="pack_seqs" id="pack_seqs">pack_seqs.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Pack the sequences such that every 32-bit word contain 16 basepairs.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/packuint8seqs.m" name="packuint8seqs" id="packuint8seqs">packuint8seqs.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Pack the sequences such that every 8-bit uint contains </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/phytree_plot.m" name="phytree_plot" id="phytree_plot">phytree_plot.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Plot a phylogenetic tree. Like Matlab's 'plot', but enable different branch colors.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/prepare_cell_multi_chipseq_pwms.m" name="prepare_cell_multi_chipseq_pwms" id="prepare_cell_multi_chipseq_pwms">prepare_cell_multi_chipseq_pwms.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Prepare some pwms which are not present in our collection but appear to</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pwm_by_name.m" name="pwm_by_name" id="pwm_by_name">pwm_by_name.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Get a pwm from a struct by it's name </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pwm_to_alphabetic.m" name="pwm_to_alphabetic" id="pwm_to_alphabetic">pwm_to_alphabetic.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Transfers the PWM into alphabetical first consensus lexicographic ..</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pwm_to_consensus.m" name="pwm_to_consensus" id="pwm_to_consensus">pwm_to_consensus.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Get the consensus sequence of a pwm</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pwm_to_iupac.m" name="pwm_to_iupac" id="pwm_to_iupac">pwm_to_iupac.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert a pwm to the closest iupac character</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pwmrcomplement.m" name="pwmrcomplement" id="pwmrcomplement">pwmrcomplement.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Get the reverse-complement of a pwm</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pwms_cut_in_half.m" name="pwms_cut_in_half" id="pwms_cut_in_half">pwms_cut_in_half.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Cut pwms to halfs and return the halfs passing some threshold</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pwms_information_content.m" name="pwms_information_content" id="pwms_information_content">pwms_information_content.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute information content for pwms vectors</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pwms_palindromic_score.m" name="pwms_palindromic_score" id="pwms_palindromic_score">pwms_palindromic_score.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Give a score for a pwm for how palindromic is it</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pwms_to_thresholds.m" name="pwms_to_thresholds" id="pwms_to_thresholds">pwms_to_thresholds.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute a 'hit' threshold for each pwm</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/pwms_to_vecs.m" name="pwms_to_vecs" id="pwms_to_vecs">pwms_to_vecs.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert 4*L matrices to vectors. </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/read_vcf.m" name="read_vcf" id="read_vcf">read_vcf.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Read Variant Call Format file to .mat</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/region_type_num2str.m" name="region_type_num2str" id="region_type_num2str">region_type_num2str.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert region type numbering coding to string</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/region_type_str2num.m" name="region_type_str2num" id="region_type_str2num">region_type_str2num.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert region type numbering coding to string</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/regions_pwm_affinity.m" name="regions_pwm_affinity" id="regions_pwm_affinity">regions_pwm_affinity.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute affinity of a DNA sequence to a pwm.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/regions_pwm_affinity_from_file.m" name="regions_pwm_affinity_from_file" id="regions_pwm_affinity_from_file">regions_pwm_affinity_from_file.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute affinity of a DNA sequence to a pwm, with input from files.</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/regions_pwm_pval_from_file.m" name="regions_pwm_pval_from_file" id="regions_pwm_pval_from_file">regions_pwm_pval_from_file.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute enrichment p-value for pwms on regions</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/reverse_complement_for_packed.m" name="reverse_complement_for_packed" id="reverse_complement_for_packed">reverse_complement_for_packed.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute the reverse complement when sequences are packed</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/run_pwms_palindromic_score.m" name="run_pwms_palindromic_score" id="run_pwms_palindromic_score">run_pwms_palindromic_score.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Give a score for a pwm for how palindromic is it</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/save_bs_mat_as_text.m" name="save_bs_mat_as_text" id="save_bs_mat_as_text">save_bs_mat_as_text.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Just take some binding sites and save them nicely as a .txt file which</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/save_expression_mat_as_text.m" name="save_expression_mat_as_text" id="save_expression_mat_as_text">save_expression_mat_as_text.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Save expression data matrix in .txt format</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/save_expression_mat_as_text_generic.m" name="save_expression_mat_as_text_generic" id="save_expression_mat_as_text_generic">save_expression_mat_as_text_generic.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Save expression data in .txt format</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/save_expression_text_as_mat.m" name="save_expression_text_as_mat" id="save_expression_text_as_mat">save_expression_text_as_mat.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Save expression data in .mat format</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/save_expression_text_as_mat_generic.m" name="save_expression_text_as_mat_generic" id="save_expression_text_as_mat_generic">save_expression_text_as_mat_generic.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Save expression data in .mat format</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/save_pwms_to_txt_file.m" name="save_pwms_to_txt_file" id="save_pwms_to_txt_file">save_pwms_to_txt_file.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Saves a set of pwms in a convenient .txt format </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/save_regions_mat_as_text.m" name="save_regions_mat_as_text" id="save_regions_mat_as_text">save_regions_mat_as_text.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Save a vector of regions as .txt file</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/seqdist.m" name="seqdist" id="seqdist">seqdist.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute the Hamming distance between two sequences. </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/sites_to_pwm.m" name="sites_to_pwm" id="sites_to_pwm">sites_to_pwm.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Generate a pwm from genomic sites</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/strand2num.m" name="strand2num" id="strand2num">strand2num.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert a strand vector marked by '+' and '-' to numeric</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/test_dna_utils.m" name="test_dna_utils" id="test_dna_utils">test_dna_utils.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>function test_dna_utils()</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/test_pwms_to_thresholds.m" name="test_pwms_to_thresholds" id="test_pwms_to_thresholds">test_pwms_to_thresholds.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Test the function pwms_to_thresholds</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/uint8seqstobinarymat.m" name="uint8seqstobinarymat" id="uint8seqstobinarymat">uint8seqstobinarymat.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Transfer the sequence into a binary matrix form. </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/unique_pwms_columns.m" name="unique_pwms_columns" id="unique_pwms_columns">unique_pwms_columns.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Get a set of pwms and get a list of all unique columns</td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/unpack_seqs.m" name="unpack_seqs" id="unpack_seqs">unpack_seqs.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Unpack sequences such that each 32-bit word contains 16 basepairs. </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/unpackuint8seqs.m" name="unpackuint8seqs" id="unpackuint8seqs">unpackuint8seqs.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Pack the sequences such that every 8-bit uint contains </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/vecs_to_pwms.m" name="vecs_to_pwms" id="vecs_to_pwms">vecs_to_pwms.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Convert 4L vectors to 4*L matrices. </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/xcorr_by_nuc.m" name="xcorr_by_nuc" id="xcorr_by_nuc">xcorr_by_nuc.m</a></td>

<td>&nbsp;-&nbsp;</td>

<td>Compute the auto-correlation at many distances </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/add_noise_to_kmers.cpp" name="add_noise_to_kmers.c" id="add_noise_to_kmers.c"><font color="003300">add_noise_to_kmers.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>nclude <stdlib.h></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/calc_all_sites_scores.cpp" name="calc_all_sites_scores.c" id="calc_all_sites_scores.c"><font color="003300">calc_all_sites_scores.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>nclude <stdlib.h></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/calc_all_sites_scores_diff_lengths.cpp" name="calc_all_sites_scores_diff_lengths.c" id="calc_all_sites_scores_diff_lengths.c"><font color="003300">calc_all_sites_scores_diff_lengths.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>nclude <stdlib.h></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/dna_utils.cpp" name="dna_utils.c" id="dna_utils.c"><font color="003300">dna_utils.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>nclude <stdlib.h></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/extract_sub_kmers.cpp" name="extract_sub_kmers.c" id="extract_sub_kmers.c"><font color="003300">extract_sub_kmers.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>nclude <stdlib.h></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/get_scores_above_threshold.cpp" name="get_scores_above_threshold.c" id="get_scores_above_threshold.c"><font color="003300">get_scores_above_threshold.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>nclude <stdlib.h></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/get_scores_above_threshold_diff_lengths.cpp" name="get_scores_above_threshold_diff_lengths.c" id="get_scores_above_threshold_diff_lengths.c"><font color="003300">get_scores_above_threshold_diff_lengths.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>nclude <stdlib.h></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/hmm_chrom_funcs.cpp" name="hmm_chrom_funcs.c" id="hmm_chrom_funcs.c"><font color="003300">hmm_chrom_funcs.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>nclude <stdlib.h></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/intervals_intersect_c.cpp" name="intervals_intersect_c.c" id="intervals_intersect_c.c"><font color="003300">intervals_intersect_c.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>nclude <stdlib.h></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/markov.cpp" name="markov.c" id="markov.c"><font color="003300">markov.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>nclude <stdlib.h></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/match_closest_vals_c.cpp" name="match_closest_vals_c.c" id="match_closest_vals_c.c"><font color="003300">match_closest_vals_c.cpp</a></td>

<td>&nbsp;-&nbsp;</td>

<td>inv_sort_perm_b = inv_perm(sort_perm_b); </td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/add_noise_to_kmers.mexa64" name="add_noise_to_kmers.mexa" id="add_noise_to_kmers.mexa"><font color="003300">add_noise_to_kmers.mexa64</a></td>

<td>&nbsp;-&nbsp;</td>

<td>LF
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/calc_all_sites_scores.mexa64" name="calc_all_sites_scores.mexa" id="calc_all_sites_scores.mexa"><font color="003300">calc_all_sites_scores.mexa64</a></td>

<td>&nbsp;-&nbsp;</td>

<td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/calc_all_sites_scores.mexw32" name="calc_all_sites_scores.mexw" id="calc_all_sites_scores.mexw"><font color="003300">calc_all_sites_scores.mexw32</a></td>

<td>&nbsp;-&nbsp;</td>

<td>�
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/calc_all_sites_scores_diff_lengths.mexa64" name="calc_all_sites_scores_diff_lengths.mexa" id="calc_all_sites_scores_diff_lengths.mexa"><font color="003300">calc_all_sites_scores_diff_lengths.mexa64</a></td>

<td>&nbsp;-&nbsp;</td>

<td>LF
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/calc_all_sites_scores_diff_lengths.mexw32" name="calc_all_sites_scores_diff_lengths.mexw" id="calc_all_sites_scores_diff_lengths.mexw"><font color="003300">calc_all_sites_scores_diff_lengths.mexw32</a></td>

<td>&nbsp;-&nbsp;</td>

<td>�
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/dna_utils.mexa64Change" name="dna_utils.mexa64Chan" id="dna_utils.mexa64Chan"><font color="003300">dna_utils.mexa64Change</a></td>

<td>&nbsp;-&nbsp;</td>

<td>LF
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/dna_utils.mexw32" name="dna_utils.mexw" id="dna_utils.mexw"><font color="003300">dna_utils.mexw32</a></td>

<td>&nbsp;-&nbsp;</td>

<td></td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/extract_sub_kmers.mexa64" name="extract_sub_kmers.mexa" id="extract_sub_kmers.mexa"><font color="003300">extract_sub_kmers.mexa64</a></td>

<td>&nbsp;-&nbsp;</td>

<td>LF
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/get_scores_above_threshold.mexa64" name="get_scores_above_threshold.mexa" id="get_scores_above_threshold.mexa"><font color="003300">get_scores_above_threshold.mexa64</a></td>

<td>&nbsp;-&nbsp;</td>

<td>LF
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/get_scores_above_threshold.mexw32" name="get_scores_above_threshold.mexw" id="get_scores_above_threshold.mexw"><font color="003300">get_scores_above_threshold.mexw32</a></td>

<td>&nbsp;-&nbsp;</td>

<td>�
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/get_scores_above_threshold_diff_lengths.mexa64" name="get_scores_above_threshold_diff_lengths.mexa" id="get_scores_above_threshold_diff_lengths.mexa"><font color="003300">get_scores_above_threshold_diff_lengths.mexa64</a></td>

<td>&nbsp;-&nbsp;</td>

<td>LF
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/get_scores_above_threshold_diff_lengths.mexw32" name="get_scores_above_threshold_diff_lengths.mexw" id="get_scores_above_threshold_diff_lengths.mexw"><font color="003300">get_scores_above_threshold_diff_lengths.mexw32</a></td>

<td>&nbsp;-&nbsp;</td>

<td>�
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/intervals_intersect_c.mexa64" name="intervals_intersect_c.mexa" id="intervals_intersect_c.mexa"><font color="003300">intervals_intersect_c.mexa64</a></td>

<td>&nbsp;-&nbsp;</td>

<td>
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/intervals_intersect_c.mexw32" name="intervals_intersect_c.mexw" id="intervals_intersect_c.mexw"><font color="003300">intervals_intersect_c.mexw32</a></td>

<td>&nbsp;-&nbsp;</td>

<td>�
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/match_closest_vals_c.mexa64" name="match_closest_vals_c.mexa" id="match_closest_vals_c.mexa"><font color="003300">match_closest_vals_c.mexa64</a></td>

<td>&nbsp;-&nbsp;</td>

<td>LF
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/match_closest_vals_c.mexw32" name="match_closest_vals_c.mexw" id="match_closest_vals_c.mexw"><font color="003300">match_closest_vals_c.mexw32</a></td>

<td>&nbsp;-&nbsp;</td>

<td>�
</tr>
<tr>
<td><a href="http://www.broadinstitute.org/~orzuk/matlab/libs/bioinfo/bioinfo.tgz" name="bioinfo" id="bioinfo"><STRONG><i>bioinfo.tgz</i></STRONG></a></td>

<td>&nbsp;-&nbsp;</td>

<td><b><i>Download the whole package. (102 .m files) </i></b></td>
</tr>
      <TR>
    </TR></TBODY></TABLE><!-- End m-file index -->
  </DIV>


  <DIV class="sep">
    <HR>
  </DIV>

  <ADDRESS>

Contact: <BR> 
&nbsp; &nbsp; O. Zuk (<A href="mailto:orzuk@broadinstitute.org">orzuk@broadinstitute.org</A>) <!-- hhmts end -->
<br> Last Modified: <!--#echo var="LAST_MODIFIED"-->  
  </ADDRESS>
</BODY></HTML>
